mempool/cat/pool.go

```diff
@@ -0,0 +1,733 @@
+	if wtx != nil {
+		return wtx.tx, true
+	}
+	return types.Tx{}, false
+}
+
+func (txmp *TxPool) IsRejectedTx(txKey types.TxKey) bool {
+	return txmp.rejectedTxCache.Has(txKey)
+}
+
+func (txmp *TxPool) WasRecentlyEvicted(txKey types.TxKey) bool {
+	return txmp.evictedTxs.Has(txKey)
+}
+
+func (txmp *TxPool) CanFitEvictedTx(txKey types.TxKey) bool {
+	info := txmp.evictedTxs.Get(txKey)
+	if info == nil {
+		return false
+	}
+	return txmp.canAddTx(info.size)
+}
+
+func (txmp *TxPool) TryReinsertEvictedTx(txKey types.TxKey, tx types.Tx, peer uint16) error {
+	info := txmp.evictedTxs.Pop(txKey)
+	if info == nil {
+		return nil
+	}
+	txmp.logger.Debug("attempting to reinsert evicted tx", "txKey", fmt.Sprintf("%X", txKey))
+	wtx := newWrappedTx(
+		tx, txKey, txmp.height, info.gasWanted, info.priority, info.sender,
+	)
+	checkTxResp := &abci.ResponseCheckTx{
+		Code:      abci.CodeTypeOK,
+		Priority:  info.priority,
+		Sender:    info.sender,
+		GasWanted: info.gasWanted,
+	}
+	return txmp.addNewTransaction(wtx, checkTxResp)
+}
+
+// CheckTx adds the given transaction to the mempool if it fits and passes the
+// application's ABCI CheckTx method. This should be viewed as the entry method for new transactions
+// into the network. In practice this happens via an RPC endpoint
+func (txmp *TxPool) CheckTx(tx types.Tx, cb func(*abci.Response), txInfo mempool.TxInfo) error {
+	// Reject transactions in excess of the configured maximum transaction size.
+	if len(tx) > txmp.config.MaxTxBytes {
+		return mempool.ErrTxTooLarge{Max: txmp.config.MaxTxBytes, Actual: len(tx)}
+	}
+
+	// This is a new transaction that we haven't seen before. Verify it against the app and attempt
+	// to add it to the transaction pool.
+	key := tx.Key()
+	rsp, err := txmp.TryAddNewTx(tx, key, txInfo)
+	if err != nil {
+		return err
+	}
+
+	// push to the broadcast queue that a new transaction is ready
+	txmp.markToBeBroadcast(key)
+
+	// call the callback if it is set
+	if cb != nil {
+		cb(&abci.Response{Value: &abci.Response_CheckTx{CheckTx: rsp}})
+	}
+	return nil
+}
+
+func (txmp *TxPool) next() <-chan types.TxKey {
+	txmp.broadcastMtx.Lock()
+	defer txmp.broadcastMtx.Unlock()
+	if len(txmp.txsToBeBroadcast) != 0 {
+		ch := make(chan types.TxKey, 1)
+		for key := range txmp.txsToBeBroadcast {
+			delete(txmp.txsToBeBroadcast, key)
+			ch <- key
+			return ch
+		}
+	}
+	return txmp.broadcastCh
+}
+
+func (txmp *TxPool) markToBeBroadcast(key types.TxKey) {
+	if !txmp.config.Broadcast {
+		return
+	}
+
+	select {
+	case txmp.broadcastCh <- key:
+	default:
+		txmp.broadcastMtx.Lock()
+		defer txmp.broadcastMtx.Unlock()
+		txmp.txsToBeBroadcast[key] = struct{}{}
+	}
+}
+
+// TryAddNewTx attempts to add a tx that has not already been seen before. It first marks it as seen
+// to avoid races with the same tx. It then call `CheckTx` so that the application can validate it.
+// If it passes `CheckTx`, the new transaction is added to the mempool solong as it has
```

Comment:
[nit] b/c I don't think solong is a word

```suggestion
// If it passes `CheckTx`, the new transaction is added to the mempool as long as it has
```